### ISSUE TITLE
Discussion: Unit: Estimated quantity

### ISSUE THREAD

**Issue by jpmckinney** [open-contracting/standard#689]
## Motivation

The [Revised GPA](https://www.wto.org/english/docs_e/legal_e/rev-gpr-94_01_e.htm) has: "the nature and the quantity of the goods or services to be procured **or, where the quantity is not known, the estimated quantity**" (emphasis added)

Presently, `quantity` is "The number of units required" and is a number. There is, to my knowledge, no way to indicate that a number is an estimate in OCDS.

We should consider a general approach to imprecision in OCDS. In terms of imprecision, we generally either know that the true value is close to a number, or between two numbers.

## Discussion

Three options are:

1. **Different fields** for precision (e.g. `quantity`, `estimatedQuantity`) and ranges (`minimumQuantity`, `maximumQuantity`)
    * Advantage:
      1. Easy to read and write a field's value.
      1. If you don't care about imprecise values, you can ignore those fields.
    * Disadvantages:
      1. Increases the number of fields in the parent object, decreasing simplicity and readability.
      1. A user may need to read all the fields to establish the value to use or display.
      1. The fields may be inconsistent, e.g. the precise/estimated value may be outside the range.
      1. The semantics for reading the values need to be established by OCDS (and implemented in both data-entry and data-use systems), e.g. if a precise value is present, then prevent data entry for the others and ignore the others for data use.
1. **Formatted values** to indicate [intervals](https://en.wikipedia.org/wiki/Interval_(mathematics)) (`[0, 100]`, etc.) or precision (e.g. `100` for precise, `~100` for estimate)
    * Advantages:
      1. Maintains the same number of fields, to maintain simplicity.
      1. A user needs to read a single field to establish the value to use or display.
      1. The field cannot be inconsistent, because it is a single value.
      1. No new semantic rules are required for data entry or data use systems.
    * Disadvantages:
      1. Requires a decoder to read a field's value if it is imprecise. (But, if you don't care about imprecise values, you can just use a programming language's number parsing and ignore exceptions.)
      1. Requires an encoder to write a field's value if it is imprecise.
1. **Objects** with fields for precision (e.g. `value`, `estimatedValue`) and ranges (e.g. `minimum`, `maximum`)
    * Advantages:
      1. Collects the fields into a single object, to mitigate complexity and maintain readability.
      1. See 1.b.
    * Disadvantages:
      1. Adds an extra layer to read and write a field's value.
      1. See 1.b-d.

## Proposal

My preference is for the second option; it's only disadvantage is requiring encoding and decoding of values. On the other hand, the first and third options (mostly the same) introduce data integrity issues (1.c) and unenforceable conformance issues (1.d); and if a user cares about imprecise values, they still need to have some logic to determine what value to use or display.

The encoding/decoding disadvantage can be addressed by authoring libraries in popular languages. On the other hand, the integrity and conformance issues are unpreventable.

I didn't discuss backwards-compatibility to avoid conflating what's possible (and ideal) with what's practical. The third option is backwards-incompatible. The first and second options are backwards-compatible: the second option relaxes quantity to be a string as well as a number.

### Notation

If we go with the second option, we will need to agree the notation.
* I propose using [ISO 31-11](https://en.wikipedia.org/wiki/ISO_31-11) for intervals; we should choose between brackets or parentheses for open intervals, or choose to support only closed intervals. 
* I propose we use tilde `~` for estimated or approximated values. The hat `^` is used in statistics to indicate an [estimated value](https://en.wikipedia.org/wiki/Hat_operator), but also used in mathematics for the [exponent operator](https://en.wikipedia.org/wiki/Caret), so it is confusing. The tilde `~` is sometimes used for [asymptotic equivalence](https://en.wikipedia.org/wiki/Approximation), but the less ambiguous `≈` is a non-ASCII character, which is more likely to run into encoding issues and is not easy to type into simple data entry systems.

**Comment by jpmckinney** [comment]
@myroslav I'd be interested in your feedback here, with respect to estimated quantity, as part of GPA extensions.

**Comment by PaulBoroday** [comment]
@jpmckinney I would suggest considering [metrics](https://github.com/open-contracting-extensions/ocds_metrics_extension) here. 

But the only thing that is missed - something like 'relatedTarget' for 'item'

**Comment by jpmckinney** [comment]
@PaulBoroday I consider the approach used by the metrics extensions to be too generic and indirect for something as basic as an item's quantity. The second option above involves a very minimal change, and only in the case where the quantity is an estimate.

**Comment by yolile** [comment]
@jpmckinney We already have, at tender level, the field "value" and "minValue", so wouldn't it be more consistent with this to also have a minimumQuantity field?

**Comment by jpmckinney** [comment]
That's option 1, and we'd need four fields total: quantity, estimatedQuantity (e.g. middle of range), minimumQuantity, maximumQuantity. See the advantages/disadvantages of each option and then let me know if you disagree with the analysis.

**Comment by jpmckinney** [comment]
@yolile I see you're talking about value from #888, not quality (this issue). I'll reply there.

**Comment by yolile** [comment]
> That's option 1, and we'd need four fields total: quantity, estimatedQuantity (e.g. middle of range), minimumQuantity, maximumQuantity. See the advantages/disadvantages of each option and then let me know if you disagree with the analysis.

I think that the option 2 is not backward compatible, can you explain why it is? and if we want to use it in Paraguay we should wait to the 1.2 version of the standard to use it, what is not possible. Also, I think that the process or rules that you have to do or write for Option 1 or 2 ar basically the same. The only real adventage of Option 2 is that you dont have to add more fields to the parent object, but, probably when you show the json in a human readable format you will show the range separately, and the same at the data input, and also if you want to calculate things with the min and max quantities you will need to separete the range. So, option 2 would require extra processing to input the max and min fields in a procurement system, and to separate the numbers in the range to show them in a report or visualization

So, I think that, at least in Paraguay the better approach would be creating an extension to add the field minQuantity, and use the quantity as the maxQuantity when the minQuantity is present, as in tender/value and tender/minValue

**Comment by jpmckinney** [comment]
Option 2 is backwards compatible because it simply relaxes the `type` of `quantity` to be number or string, instead of just a number. For the option to be backwards-incompatible, it would need to add restrictions.

Option 2 has additional advantages. As noted in the proposal, options 1 and 3 introduce:

* Data integrity issues (1.c): The fields can be inconsistent, e.g. the `quantity` and/or `estimatedQuantity` values can be outside the `minimumQuantity`-`maximumQuantity` range.
* Unenforceable conformance issues (1.d): The semantics for reading the values need to be established by OCDS (and implemented in both data-entry and data-use systems), for example: if the precise `quantity` is present, then prevent data entry of the other fields, and ignore the other fields for data use.

If, like in Paraguay, all you want is minQuantity, then option 1 is simplest, but this issue isn't just about minimum quantities but also about estimated quantities.

**Comment by yolile** [comment]
> Option 2 is backwards compatible because it simply relaxes the type of quantity to be number or string, instead of just a number. For the option to be backwards-incompatible, it would need to add restrictions.

Aha, but I think it will break any existing tool that expects `quantity` to be a number and do calculations with it. Maybe another mixed solution could be to leave the quantity field as it is and add just one other field, `estimatedQuantity` that can be an interval or an approximated value?

**Comment by jpmckinney** [comment]
True. Our target is backwards-compatibility, i.e. old versions of data are still valid input for new tools (as long as tools don't drop support for deprecated features). We don't have a target for forwards-compatibility, i.e. new versions of data are still valid input for old tools. For example, any tool that did analysis of parties beyond `id` and `name` was broken by 1.1. We can consider forwards-compatibility in evaluating options, but it's not a strict requirement.

**Comment by yolile** [comment]
> For example, any tool that did analysis of parties beyond id and name was broken by 1.1

Not sure if that case is the same case because 1) 1.1 only deprecated the other fields and 2) as none of the removed fields were required ones the tools needs to ask if the field exists first anyway (so true, the tool won't work anymore but it won't crash either)

But agree that option 2 is the better one. @jpmckinney To mark this issue as ready to PR, should then the quantity field now:
1. Change its type to:
```json
 "type": [
            "number",
            "string",
            "null"
          ]
```
2. Have a ` pattern` field, accordingly to the proposed [notation](https://github.com/open-contracting/standard/issues/689#issue-315700406)
3. Update its description to reflect the notation as well

?

**Comment by jpmckinney** [comment]
That is the proposal, yes. However, given this change might be disruptive, I think it important to first validate the demand. As such, I think the next step for this issue is for it to be raised in a community call for further input.

In terms of internal coordination, in general, any issue with the label "discussion" first needs to have that label removed, before we can progress to a PR. In this case, it retains the label because of the need for further input.

**Comment by duncandewhurst** [comment]
I agree that we should validate the demand before progressing with Option 2.

We should also consider the impact on users of tabular OCDS data, either in SQL databases or spreadsheets, for whom using a programming language's number parser and ignoring errors isn't an option.

**Comment by jpmckinney** [comment]
For framework agreements, we added `awards.estimatedValue`, `awards.maximumValue` and `tender.maximumValue` (to the existing `awards.value`, `tender.value` and `tender.minValue`.

So, to be consistent, we would go with option 1 here.

Note that, at the tender stage, it has been clarified in the field descriptions that all values (whether min, max or plain) are estimated. (Hence, no `tender.estimatedValue`).

Similarly, the GPA is probably overly verbose to be clear that values in tender notices can be estimates. That said, I don't think these estimated quantities are ever more than one number.

So, there is nothing to do – plus there is no demand. Closing.
